dtfs-central-api/src/helpers/get-fee-record-payment-entity-group-reconciliation-data.ts

```diff
@@ -0,0 +1,59 @@
+import { orderBy } from 'lodash';
+import { FEE_RECORD_STATUS } from '@ukef/dtfs2-common';
+import { TfmUsersRepo } from '../repositories/tfm-users-repo';
+import { FeeRecordReconciledByUser } from '../types/utilisation-reports';
+import { FeeRecordPaymentEntityGroup } from '../types/fee-record-payment-entity-group';
+
+/**
+ * Gets the reconciliation data for a fee record payment entity group
+ * @param group - The fee record payment entity group
+ * @returns The group reconciliation data
+ * @throws {Error} If the groups fee records array is empty
+ * @throws {Error} If any of the fee records at the `RECONCILED` status have a null `dateReconciled` property
+ */
+export const getFeeRecordPaymentEntityGroupReconciliationData = async (
+  group: FeeRecordPaymentEntityGroup,
+): Promise<{
+  reconciledByUser?: FeeRecordReconciledByUser;
+  dateReconciled?: Date;
+}> => {
+  if (group.feeRecords.length === 0) {
+    throw new Error('Fee record payment entity group cannot have an empty fee records array');
+  }
+
+  const reconciledFeeRecords = group.feeRecords.filter(({ status }) => status === FEE_RECORD_STATUS.RECONCILED);
+
+  if (reconciledFeeRecords.length === 0) {
+    return {};
+  }
+
+  if (reconciledFeeRecords.some(({ dateReconciled }) => dateReconciled === null)) {
+    throw new Error(`Fee records at the '${FEE_RECORD_STATUS.RECONCILED}' status cannot have a null 'dateReconciled' property`);
+  }
+
+  const feeRecordsSortedByDateReconciledDescending = orderBy(reconciledFeeRecords, [(feeRecord) => feeRecord.dateReconciled!.getTime()], ['desc']);
+  const mostRecentlyReconciledFeeRecord = feeRecordsSortedByDateReconciledDescending.at(0);
+  if (!mostRecentlyReconciledFeeRecord) {
+    // Based on all the above checks, this error should never be reached, hence not being documented
+    throw new Error('Something went wrong when sorting the fee records to get the most recently reconciled fee record');
```

Comment:
has this error been tested? apologies if I have missed it in the test file

Reply:
No it hasn't. It is an error that should be impossible to reach based on all the previous checks, and so it's not actually possible to even test it. It is just a guard clause introduced to work with typescript

Reply:
WIthout diving deep into this - @RLavender98 , if this is an impossible scenario as Josh mentions - do we even need this? Otherwise is it possible to add a test?

If it's not possible to replicate with a unit test, I don't think we should throw an error here. Maybe a `console.error` here would be better for future debugging purposes?

Reply:
On line 26 we verify that the reconciledFeeRecords have non zero length. This should give us carried forwards type safety allowing us to get the first element of the array and know that it exists.

However, the return type of `orderBy` by here results in `FeeRecordEntity[]` and a loss of that type safety because it doesn't include in it's typing an assertion that the result will have the same length as the input even though we know it will. Hence, the IDE and compiler are not aware that `const mostRecentlyReconciledFeeRecord = feeRecordsSortedByDateReconciledDescending.at(0)` will definitely give us a defined fee record and so, without this check we get a ts error on line 43 when we try to deconstruct.

I can instead do `const mostRecentlyReconciledFeeRecord = feeRecordsSortedByDateReconciledDescending.at(0)!;` and then not have the check which should I think be the same I guess so I will make that change.